### PR TITLE
Add environment scope to gitlab_group_variable

### LINKED ISSUE #610: Support group level CI/CD variable environment scope
### Terraform Version
```
terraform -v
Terraform v0.14.10

Your version of Terraform is out of date! The latest version
is 0.15.0. You can update by downloading from https://www.terraform.io/downloads.html
```

### Affected Resource(s)
Please list the resources as a list, for example:
- gitlab_group_variable (data source and resource)

### Terraform Configuration Files
```hcl
resource "gitlab_group" "example" {
  name = "example"
  path = "example"
}

resource "gitlab_group_variable" "example" {
  group  = gitlab_group.example.id
  key    = "example"
  value  = "example"

  environment_scope = "staging"
}
```

### Debug Output
N/A.

### Panic Output
N/A.

### Expected Behavior
Ability to create a group level variable and set the environment_scope field.

### Actual Behavior
Not yet available :)

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

### Important Factoids
Only supported in GitLab Premium and above.

### References
- Depends on https://github.com/xanzy/go-gitlab/pull/1137 to be released

### PR DESCRIPTION
**Pull request by njdart**
Add support for `environment_scope` to group level variables

Fixes #610
Supersedes #611

### CODE REVIEW

**Comment by mkjmdski** [comment]
Could you add test for that? It should be enough if you set `*` environment scope in `testAccGitlabGroupVariableConfig` and later change it in `testAccGitlabGroupVariableUpdateConfig`.

**Comment by njdart** [comment]
@mkjmdski Thanks for getting back

I don't think I can add a meaningful test here since scoped group-level variables is a premium feature (as per https://docs.gitlab.com/ee/ci/environments/#scope-environments-with-specs and the linked issue on addition of group-level scope https://gitlab.com/gitlab-org/gitlab/-/issues/2874).

I tried making the changes you suggested but the test fails with `Error destroying resource! WARNING: Dangling resources.` ... `errors during refresh: GET http://127.0.0.1:8080/api/v4/groups/394/variables/key_10d0d: 404 {message: 404 GroupVariable Not Found}`. I think this is because modifying the scope of a variable in the (FOSS?) test environment is a no-op and sets the scope back to `*`, but still returns 200. Teardown then fails to find the variable with the new scope because it wasn't correctly changed. Invoking the API manually shows the following:

```
[nic@ARCHY] ~ curl -s 'http://127.0.0.1:8080/api/v4/groups/394/variables/key_10d0d?access_token=ACCTEST1234567890123' | jq
{
  "variable_type": "file",
  "key": "key_10d0d",
  "value": "value-10d0d",
  "protected": false,
  "masked": false,
  "environment_scope": "*"
}
```

```
[nic@ARCHY] ~ curl -s 'http://127.0.0.1:8080/api/v4/groups/394/variables/key_10d0d?access_token=ACCTEST1234567890123' -X PUT -H 'Content-Type: application/json' -d '{"environment_scope": "review/*"}' | jq
{
  "variable_type": "file",
  "key": "key_10d0d",
  "value": "value-10d0d",
  "protected": false,
  "masked": false,
  "environment_scope": "*"
}
```

What are your thoughts? I've deployed a build version of this in CI (using developer overrides) and it's been working fine so far, just not sure how to test this

**Comment by mkjmdski** [comment]
Hi @njdart 

As far as I remember the github actions in here are supporting running test in EE edition, you can duplicate the normal tests which are run in here and provide to them this option: `SkipFunc: isRunningInCE`, you can find examples on that in `gitlab_project` tests. However I am not a maintainer of that repo so maybe it's just fine to add this feature without tests.

**Comment by njdart** [comment]
OK, I've added a step in the tests to check modifying the environment scope with the flag you mentioned. I haven't been able to test it locally though. Can the tests be enabled for this MR please?

I've also updated the docs for `environment_scope` to mention limitations for non-EE instances

**Comment by njdart** [comment]
@mkjmdski Can this be progressed?

**Comment by mkjmdski** [comment]
@roidelapluie could you enable pipelines in here?

File: gitlab/resource_gitlab_group_variable.go
Comment (njdart) [reply gitlab/resource_gitlab_group_variable.go:16]: @armsnyder (or @mkjmdski) can this be re-reviewed please?

**Comment by christoph-kluge** [comment]
Is there anything missing or any support needed to move this PR forward?

**Comment by deepbrook** [comment]
@armsnyder what's missing here?

**Review by timofurrer — CHANGES_REQUESTED** [review]
LGTM 🍰 

Please rebase to latest master and solve the conflicts. Otherwise, I'd like to get this merged.

@armsnyder @nagyv

File: gitlab/resource_gitlab_group_variable.go
Comment (timofurrer) [inline gitlab/resource_gitlab_group_variable.go:19]: Please use the context from the CRUD functions with `req.WithContext` here.

File: gitlab/resource_gitlab_group_variable.go
Comment (armsnyder) [reply gitlab/resource_gitlab_group_variable.go:16]: Thanks for testing the upgrade 👏

**Review by armsnyder — APPROVED** [review]
lgtm. Please rebase if you can.

File: gitlab/resource_gitlab_group_variable.go
Comment (njdart) [reply gitlab/resource_gitlab_group_variable.go:21]: I've updated the group_variables TF (258ceb8) definition [in accorance with the Terraform SDK v2 guide](https://www.terraform.io/plugin/sdkv2/guides/v2-upgrade-guide#more-support-for-context-context) to use the CRUD-with-context functions, and am now passing the context object into the gitlab HTTP requests (57f5be6).

I couldn't see this being done elsewhere in the project, so let me know if I've missed something and am doing it wrong.

Comment (timofurrer) [reply gitlab/resource_gitlab_group_variable.go:21]: There is an open MR #791 introduced this to the entire code base.
Please rebase, and lgtm 🍰

Comment (njdart) [reply gitlab/resource_gitlab_group_variable.go:21]: I have rebased against master. That MR looks like it might be merged relatively soon, if so should I wait and rebase again once it's in?

Comment (njdart) [reply gitlab/resource_gitlab_group_variable.go:21]: I've rebased against master and tidied up now #791 is merged. I should have also fixed the linting error that was failing earlier

**Comment by IrmantasMarozas** [comment]
Can't wait! Didn't read the doc properly and now wasted quite a bit of time expecting group to act same as project 👍

File: gitlab/resource_gitlab_group_variable.go
Comment (nagyv) [reply gitlab/resource_gitlab_group_variable.go:21]: @timofurrer Please, resolve this thread if you are ok with the current state. I will merge once you approve.

Comment (timofurrer) [reply gitlab/resource_gitlab_group_variable.go:21]: I don't know why, but I can't resolve it :'(

**Review by timofurrer — APPROVED** [review]

File: gitlab/resource_gitlab_group_variable.go
Comment (njdart) [reply gitlab/resource_gitlab_group_variable.go:21]: I have a resolve button... I'm happy we've resolved this conversation, but was wandering if the acceptance-ee test suite should be run before this gets merged?

**Comment by nagyv** [comment]
As https://github.com/gitlabhq/terraform-provider-gitlab/pull/802 might cause conflicts, but is super valuable, I will keep merging this until https://github.com/gitlabhq/terraform-provider-gitlab/pull/802 is merged. 

If someone has time today to fix https://github.com/gitlabhq/terraform-provider-gitlab/pull/802 that would be great. Unfortunately, I broke the pipeline there as I tried to fix a conflict but won't have time to fix it.

**Comment by armsnyder** [comment]
@nagyv I'm ok with fixing conflicts and merging #802 last, after the rest of milestone PRs get in.

**Pull request merged**
